sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTranspose.scala

```diff
@@ -0,0 +1,164 @@
+
+    // Example of the original DataFrame:
+    // +---+-----+-----+
+    // | id|col1 |col2 |
+    // +---+-----+-----+
+    // |  1|  10 |  20 |
+    // |  2|  30 |  40 |
+    // +---+-----+-----+
+    //
+    // After transposition, the finalMatrix will look like:
+    // [
+    //   ["col1", 10, 30],  // Transposed row for col1
+    //   ["col2", 20, 40]   // Transposed row for col2
+    // ]
+
+    for (i <- 0 until numTransposedRows) {
+      // Insert non-index column name as the first element in each transposed row
+      finalMatrix(i)(0) = UTF8String.fromString(nonIndexColumnNames(i))
+
+      for (j <- 1 until numTransposedCols) {
+        // Insert the transposed data
+
+        // Example: If j = 2, then row = fullCollectedRows(1)
+        // This corresponds to the second row of the original DataFrame: InternalRow(2, 30, 40)
+        val row = fullCollectedRows(j - 1)
+
+        // Example: If i = 0 (for "col1"), and j = 2,
+        // then finalMatrix(0)(2) corresponds to row.get(1, nonIndexColumnDataTypes(0)),
+        // which accesses the value 30 from InternalRow(2, 30, 40)
+        finalMatrix(i)(j) = row.get(i + 1, nonIndexColumnDataTypes(i))
+      }
+    }
+    finalMatrix
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+    _.containsPattern(TreePattern.UNRESOLVED_TRANSPOSE)) {
+    case t @ UnresolvedTranspose(indexColumn, child) if indexColumn.resolved && child.resolved =>
+
+      // Cast the index column to StringType
+      val indexColumnAsString = indexColumn match {
+        case attr: Attribute if attr.dataType.isInstanceOf[AtomicType] =>
+          Alias(Cast(attr, StringType), attr.name)()
+        case attr: Attribute =>
+          throw new IllegalArgumentException(
+            s"Index column must be of atomic type, but found: ${attr.dataType}")
+        case _ =>
+          throw new IllegalArgumentException(
+            s"Index column must be an atomic attribute")
+      }
+
+      // Cast non-index columns to the least common type
+      val nonIndexColumnsAttr = child.output.filterNot(
+        _.exprId == indexColumn.asInstanceOf[Attribute].exprId)
+      val nonIndexTypes = nonIndexColumnsAttr.map(_.dataType)
+      val commonType = leastCommonType(nonIndexTypes)
+      val nonIndexColumnsAsLCT = nonIndexColumnsAttr.map { attr =>
+        Alias(Cast(attr, commonType), attr.name)()
+      }
+
+      // Sort by index column values, and collect the casted frame
+      val allCastCols = indexColumnAsString +: nonIndexColumnsAsLCT
+      val sortedChild = Sort(
+        Seq(SortOrder(indexColumn.asInstanceOf[Attribute], Ascending)),
+        global = true,
+        child
+      )
+      val projectAllCastCols = Project(allCastCols, sortedChild)
+      val queryExecution = sparkSession.sessionState.executePlan(projectAllCastCols)
+      val fullCollectedRows = queryExecution.executedPlan.executeCollect()
```

Comment:
shall we add a LIMIT to avoid collecting too many data?

Reply:
we can add LIMIT dataFrameTransposeMaxValues + 1

Reply:
Good idea. Added.